packages/protocol/contracts/governance/VestingInstance.sol

```diff
@@ -0,0 +1,332 @@
+    paused = true;
+    pauseEndTime = block.timestamp.add(pausePeriod);
+    emit WithdrawalPaused(pausePeriod, block.timestamp);
+  }
+
+  /**
+   * @dev Calculates the amount that has already vested but hasn't been withdrawn yet.
+   * @param timestamp the timestamp at which to calculate the withdrawable amount
+   * @return The withdrawable amount at the timestamp
+   * @dev Function is also made public in order to be called for informational purpose
+   */
+  function getWithdrawableAmountAtTimestamp(uint256 timestamp) public view returns (uint256) {
+    return calculateVestedAmountAtTimestamp(timestamp).sub(currentlyWithdrawn);
+  }
+
+  /**
+   * @notice Calculates both non-locked and locked gold balance parts and returns their sum as the total balance of the vesting instance
+   * @return The total vesting instance balance
+   */
+  function getVestingInstanceTotalBalance() public view returns (uint256) {
+    uint256 nonLockedBalance = getGoldToken().balanceOf(address(this));
+    uint256 lockedBalance = getLockedGold().getAccountTotalLockedGold(address(this));
+    return nonLockedBalance.add(lockedBalance);
+  }
+
+  /**
+   * @dev Calculates the amount that has already vested.
+   * @param timestamp the timestamp at which to calculate the already vested amount
+   */
+  function calculateVestedAmountAtTimestamp(uint256 timestamp) private view returns (uint256) {
+    uint256 currentBalance = getVestingInstanceTotalBalance();
+    uint256 totalBalance = currentBalance.add(currentlyWithdrawn);
+    uint256 vestingPeriods = vestingSchedule.vestingAmount.div(vestingSchedule.vestAmountPerPeriod);
+
+    if (timestamp < vestingSchedule.vestingCliffStartTime) {
+      return 0;
+    }
+    if (
+      timestamp >=
+      vestingSchedule.vestingStartTime.add(vestingPeriods.mul(vestingSchedule.vestingPeriodSec)) ||
+      revoked
+    ) {
+      return totalBalance;
+    }
+
+    uint256 timeSinceStart = timestamp.sub(vestingSchedule.vestingStartTime);
+    uint256 timeToEnd = vestingPeriods.mul(vestingSchedule.vestingPeriodSec);
+    return totalBalance.mul(timeSinceStart).div(timeToEnd);
+  }
+
+  /**
+   * @notice A wrapper function for the lock gold method
+   * @param value the value of gold to be locked
+   * @dev To be called only by the beneficiary of the vesting
+   */
+  function lockGold(uint256 value) external nonReentrant onlyBeneficiary {
+    // the beneficiary may not lock more than the vesting currently has available
+    require(
+      value <= getGoldToken().balanceOf(address(this)),
+      "Gold amount to lock is greater than the currently vested amount"
+    );
+    getLockedGold().lock.gas(gasleft()).value(value)();
+  }
+
+  /**
+   * @notice A wrapper function for the unlock gold method function
+   * @param value the value of gold to be unlocked for the vesting instance
+   * @dev To be called only by the beneficiary of the vesting
+   */
+  function unlockGold(uint256 value) external nonReentrant onlyBeneficiary {
```

Comment:
An elegant way (and future proof) of implementing all this call is to implement some type of dynamic proxy.

Image we expose a function `sendTx(to:address, data: bytes)`. and then you can implement it as `to.send(data)`. Since this is  dangerous, you should have a whitelist of `to` addresses that you can call, and for each address also have a whitelist of methods you can call on each (basically the first 4 bytes of `data`).

This is not required for now, but imagine a future where LockedGold or Accounts ABI change,then you wouldn't be able to call the methods anymore. you'll need to change all the vestingInstances